ntp-daemon/src/system.rs

```diff
@@ -0,0 +1,184 @@
+            clock_select.system_jitter,
+            clock_select.system_peer_snapshot.root_delay,
+            clock_select.system_peer_snapshot.root_dispersion,
+            clock_select.system_peer_snapshot.leap_indicator,
+            clock_select.system_peer_snapshot.time,
+        );
+
+        // Handle situations needing extra processing
+        match adjust_type {
+            ClockUpdateResult::Panic => {
+                panic!(
+                    r"Unusually large clock step suggested,
+                            please manually verify system clock and reference clock 
+                                 state and restart if appropriate."
+                )
+            }
+            ClockUpdateResult::Step => {
+                peers.reset_all();
+
+                reset_epoch = reset_epoch.inc();
+                reset_tx.send_replace(reset_epoch);
+            }
+            _ => {}
+        }
+
+        // Handle updating system snapshot
+        if let ClockUpdateResult::Ignore = adjust_type {
+            // ignore this update
+        } else {
+            let mut global = global_system_snapshot.write().await;
+            global.poll_interval = controller.preferred_poll_interval();
+            global.leap_indicator = clock_select.system_peer_snapshot.leap_indicator;
+        }
+    }
+
+    // the channel closed and has no more messages in it
+    Ok(())
+}
+
+#[derive(Debug, Clone, Copy)]
+enum PeerStatus {
+    Demobilized,
+    AwaitingReset,
+    Valid(PeerSnapshot),
+}
+
+struct Peers {
+    peers: Box<[PeerStatus]>,
+}
+
+impl Peers {
+    fn new(length: usize) -> Self {
+        Self {
+            peers: vec![PeerStatus::AwaitingReset; length].into(),
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.peers.len()
+    }
+
+    fn valid_snapshots(&self) -> impl Iterator<Item = PeerSnapshot> + '_ {
+        self.peers
+            .iter()
+            .filter_map(|peer_status| match peer_status {
+                PeerStatus::Demobilized | PeerStatus::AwaitingReset => None,
+                PeerStatus::Valid(snapshot) => Some(*snapshot),
+            })
+    }
+
+    fn receive_update(&mut self, msg: MsgForSystem, current_reset_epoch: ResetEpoch) {
+        match msg {
+            MsgForSystem::MustDemobilize(index) => {
+                self.peers[index.index] = PeerStatus::Demobilized;
+            }
+            MsgForSystem::Snapshot(index, msg_reset_epoch, snapshot) => {
+                if current_reset_epoch == msg_reset_epoch {
+                    self.peers[index.index] = PeerStatus::Valid(snapshot);
+                }
+            }
+        }
+    }
+
+    fn reset_all(&mut self) {
+        for peer_status in self.peers.iter_mut() {
+            *peer_status = match peer_status {
+                PeerStatus::Demobilized => PeerStatus::Demobilized,
+                PeerStatus::AwaitingReset => PeerStatus::AwaitingReset,
+                PeerStatus::Valid(_) => PeerStatus::AwaitingReset,
```

Comment:
Minor nit, although this helps more when the nodata state is also there: I would suggest using the _ matcher here as this really is just set to awaitingreset unless we are/have demobilized.